pkg/controller/dnsendpoint/dnsendpoint_controller.go

```diff
@@ -0,0 +1,217 @@
+package dnsendpoint
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+	"github.com/openshift/hive/pkg/constants"
+	"github.com/openshift/hive/pkg/controller/dnsendpoint/nameserver"
+	hivemetrics "github.com/openshift/hive/pkg/controller/metrics"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	"github.com/openshift/hive/pkg/manageddns"
+)
+
+const (
+	controllerName = "dnsendpoint"
+)
+
+// Add creates a new DNSZone Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	logger := log.WithField("controller", controllerName)
+	c := controllerutils.NewClientWithMetricsOrDie(mgr, controllerName)
+
+	credsSecretName := os.Getenv(constants.ExternalDNSAWSCredsEnvVar)
+	if credsSecretName == "" {
+		logger.Info("no aws creds for external DNS; controller disabled")
+		return nil
+	}
+	logger.Infof("using aws creds for external DNS stored in %q secret", credsSecretName)
+
+	managedDomains, err := manageddns.ReadManagedDomainsFile()
```

Comment:
This is the tactic that we take with all of the configuration from HiveConfig. A change in HiveConfig causes a restart of the controllers pod. The controllers are not allowed to read from HiveConfig directly.

As long as the changes to the list of managed domains is additive, I do not see any problem. There are wrinkles when a managed domain is removed. If there is a clusterdeployment with managed DNS that is using a root domain that is no longer in the list of managed domains, I am not sure what behavior Hive should take. The way the code here is written, the NS records will continue to exist and no changes will be made to them, including deleting them when the clusterdeployment is deleted. In fact, the clusterdeployment will never be cleaned up since the finalizer on the DNSEndpoint will not be removed.

Reply:
Would you like to incorporate this in  V1?

Reply:
Ideally, yes. But it is low priority.